src/main/kotlin/uk/gov/gdx/datashare/service/DataReceiverService.kt

```diff
@@ -33,25 +34,37 @@ class DataReceiverService(
   suspend fun sendToDataProcessor(eventPayload: EventToPublish) {
 
     // check if client is allowed to send
-    val dataProvider = dataProviderRepository.findById(authenticationFacade.getUsername())
-      ?: throw RuntimeException("Client ${authenticationFacade.getUsername()} is not a known client")
+    val subscription = publisherSubscriptionRepository.findByClientIdAndEventType(
+      authenticationFacade.getUsername(),
+      eventPayload.eventType.toString()
+    ) ?: throw RuntimeException("Publisher ${authenticationFacade.getUsername()} does not have a subscription for event type ${eventPayload.eventType}")
 
-    // check if client is allowed to send this type of event
-    if (dataProvider.eventType != eventPayload.eventType.toString()) {
-      throw RuntimeException("Client ${dataProvider.clientName} is not allowed to provide ${eventPayload.eventType} events")
+    val dataSet = eventDatasetRepository.findById(subscription.datasetId)
+      ?: throw RuntimeException("Client ${authenticationFacade.getUsername()} is not a known dataset")
```

Comment:
This message doesn't quite tie together, I'd also be curious if there's any risk of leaking more data than we should through here